spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -318,9 +332,30 @@ protected void doStop() {
 
 		private final AckMode ackMode = getAckMode();
 
+		private final boolean isManualAck = this.ackMode.equals(AckMode.MANUAL);
+
+		private final boolean isManualImmediateAck = this.ackMode.equals(AckMode.MANUAL_IMMEDIATE)
+				|| this.ackMode.equals(AckMode.MANUAL_IMMEDIATE_SYNC);
+
+		private final boolean isAnyManualAck = this.isManualAck || this.isManualImmediateAck;
+
+		private final boolean isRecordAck = this.ackMode.equals(AckMode.RECORD);
+
 		private final boolean syncCommits = KafkaMessageListenerContainer.this.syncCommits;
 
-		private Thread consumerThread;
+		private final long pauseAfter = getPauseAfter();
+
+		private final boolean pauseEnabled = isPauseEnabled();
+
+		private final BlockingQueue<ConsumerRecords<K, V>> recordsToProcess = new LinkedBlockingQueue<>(getQueueDepth());
+
+		private final int queueDepth = getQueueDepth();
+
+		private final BlockingQueue<ConsumerRecord<K, V>> acks = new LinkedBlockingQueue<>();
```

Comment:
I've thought about that, too - and it seems more efficient as it doesn't require the application to re-fetch already retrieved messages. But there are two reasons why I believe this is unsafe:  

a) `onPartitionsRevoked()` does not provide the list of revoked partitions, but the list of currently allocated partitions (as counterintuitive as it sounds) - see https://kafka.apache.org/090/javadoc/org/apache/kafka/clients/consumer/ConsumerRebalanceListener.html#onPartitionsRevoked(java.util.Collection), so it is not safe to continue processing locally stored messages - as we don't know which of them are still our responsibility. If we ignored `onPartitionsRevoked()` and relied only on `onPartitionsAssigned()`, we had no guarantee in which order do clients process the `onPartitionsAssigned()` calls, so it is perfectly possible that a client that loses partitions is notified of that only after another client has already started fetching from the same partitions - so we need to work based on the contract of `onPartitionsRevoked()` : the API guarantees that all calls to `onPartitionsRevoked()` (where we can bring all clients to a stable state) will be handled before any of the clients receive a `onPartitionsAssigned()` (where we can resume from a stable state). 

b) Even if we were able to somehow deal with a) correctly I think that there is at least one scenario where simple filtering can result in race conditions: considering a client that starts with all partitions, and another client that joins temporarily and leaves. The initial client will start filtering out messages, but when the other client leaves, it will regain all partitions again. If it didn't get to filter out all the messages from the previously lost partitions (that it has queued), it will start processing the leftover messages (as they now match the updated partition list)  - at a pretty arbitrary point - at the risk of either duplicating the processing done by the other client, or dropping messages, or processing them in the wrong order. It's not a very typical scenario, but points out why filtering would be unsafe.

So I think that being more conservative in the case of rebalancing would pay off, even if we were to re-fetch data in some clients - I think we should treat rebalancing as an infrequent event, and  allow it to be more expensive.  What we could do is to try to give a consumer that loses partitions a chance to complete its work, before returning from `onPartitionsRevoked()`.